File: 2025/Day06/Solver.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../helper/FileReader.h"

typedef unsigned long long ul;

int main(int argc, char *argv[]) {
    const char *inputFile = (argc > 1) ? argv[1] : "Day6/in.txt";
    
    FileContent lines = readLines(inputFile);

    ul p1 = 0;
    ul p2 = 0;

    int idx = 0;
    int R = lines.count;
    int C = strlen(lines.content[0]);
    while (idx<C){
        int start = idx;
        while (idx+1==C || idx+1<C && lines.content[R-1][idx+1] == ' ')idx++;
        char op = lines.content[R-1][start];

        ul numP1 = op=='*'?1:0;
        for (int i=0; i<R-1; i++){
            ul curr = 0;
            for (int j=start; j<idx; j++){
                char c = lines.content[i][j];
                if (c != ' ') curr = curr * 10 + (c-'0');
            }
            if (op=='*') numP1*=curr;
            else numP1+=curr;
        }

        ul numP2 = op=='*'?1:0;
        for (int j=start; j<idx; j++){
            ul curr = 0;
            for (int i=0; i<R-1; i++){
                char c = lines.content[i][j];
                if (c != ' ') curr = curr * 10 + (c-'0');
            }
            if (op=='*') numP2*=curr;
            else numP2+=curr;
        }
        p1+=numP1;
        p2+=numP2;
        idx++;
    }
    
    freeFileLines(&lines);

    printf("Part 1: %llu\n", p1);
    printf("Part 2: %llu\n", p2);
    
    return 0;
}

File: 2025/Day06/Solver.java
package Day06;
import helper.FileReader;
import java.util.*;

public class Solver{
    public static long calcSolution(int startBlock, int endBlock, String[] lines, boolean vertical){
        char op = lines[lines.length-1].charAt(startBlock);
        long num = op=='*'?1:0;

        int startA = vertical ? 0 : startBlock;
        int endA = vertical ? lines.length-1 : endBlock;
        int startB = vertical ? startBlock : 0;
        int endB = vertical ? endBlock : lines.length-1;

        for (int i=startA; i<endA; i++){
            long curr = 0;
            for (int j=startB; j<endB; j++){
                char c = vertical?lines[i].charAt(j):lines[j].charAt(i);
                if (c != ' ') curr = curr * 10 + (c-'0');
            }
            if (op=='*') num*=curr;
            else num+=curr;
        }
        return num;
    }

    public static void main(String[] args) throws Exception {
        FileReader fr = new FileReader(args);
        String[] lines = fr.readLines();

        long p1 = 0;
        long p2 = 0;

        int idx = 0;
        int R = lines.length;
        int C = lines[0].length();
        while (idx<C){
            int start = idx;
            while (idx+1==C || idx+1<C && lines[R-1].charAt(idx+1) == ' ')idx++;
            p1 += calcSolution(start, idx, lines, true);
            p2 += calcSolution(start, idx, lines, false);
            idx++;
        }

        System.out.println("Part 1: "+ p1);
        System.out.println("Part 2: "+ p2);
    }
}